From 812fe245d90e0735efedb4f747ab7ade3c3a9089 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:53:24 -0400
Subject: [tools] perf: Fix strict alias issue for find_first_bit

Message-id: <1408953210-26343-317-git-send-email-jolsa@redhat.com>
Patchwork-id: 88939
O-Subject: [PATCH RHEL7.1 BZ1133083 316/322] perf tools: Fix strict alias issue for find_first_bit
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit b39c2a57a00a841f057a75b41df4c26173288b66
Author: Jiri Olsa <jolsa@redhat.com>
Date: Wed Feb 26 18:14:26 2014 +0100

description
===========
When compiling perf tool code with gcc 4.4.7 I'm getting
following error:

    CC       util/session.o
  cc1: warnings being treated as errors
  util/session.c: In function 'perf_session_deliver_event':
  tools/perf/util/include/linux/bitops.h:109: error: dereferencing pointer 'p' does break strict-aliasing rules
  tools/perf/util/include/linux/bitops.h:101: error: dereferencing pointer 'p' does break strict-aliasing rules
  util/session.c:697: note: initialized from here
  tools/perf/util/include/linux/bitops.h:101: note: initialized from here
  make[1]: *** [util/session.o] Error 1
  make: *** [util/session.o] Error 2

The aliased types here are u64 and unsigned long pointers, which is safe
for the find_first_bit processing.

This error shows up for me only for gcc 4.4 on 32bit x86, even for
-Wstrict-aliasing=3, while newer gcc are quiet and scream here for
-Wstrict-aliasing={2,1}. Looks like newer gcc changed the rules for
strict alias warnings.

The gcc documentation offers workaround for valid aliasing by using
__may_alias__ attribute:

  http://gcc.gnu.org/onlinedocs/gcc-4.4.0/gcc/Type-Attributes.html

Using this workaround for the find_first_bit function.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/include/linux/bitops.h b/tools/perf/util/include/linux/bitops.h
index 45cf10a..dadfa7e 100644
--- a/tools/perf/util/include/linux/bitops.h
+++ b/tools/perf/util/include/linux/bitops.h
@@ -87,13 +87,15 @@ static __always_inline unsigned long __ffs(unsigned long word)
  return num;
 }
 
+typedef const unsigned long __attribute__((__may_alias__)) long_alias_t;
+
 /*
  * Find the first set bit in a memory region.
  */
 static inline unsigned long
 find_first_bit(const unsigned long *addr, unsigned long size)
 {
- const unsigned long *p = addr;
+ long_alias_t *p = (long_alias_t *) addr;
  unsigned long result = 0;
  unsigned long tmp;
 
-- 
1.7.1